wayland-scanner: split from wayland

alyssais opened the pull request:
###### Description of changes

We've used a `wayland-scanner = wayland.bin` alias for ages, to make packages clearer and allow them to be independently overridden. Going the whole way into splitting them into separate packages is useful because it means we can have different meta.platforms attributes for libwayland and wayland-scanner.

There is no duplication in outputs between the two packages — they don't install any files in common.

Draft because I'm going to need to test the downstream packages and make sure none were relying on pulling wayland-scanner from the wayland attribute.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alyssais committed:
wayland-scanner: split from wayland

We've used a wayland-scanner = wayland.bin alias for ages, to make
packages clearer and allow them to be independently overridden.
Going the whole way into splitting them into separate packages is
useful because it means we can have different meta.platforms
attributes for libwayland and wayland-scanner.

There is no duplication in outputs between the two packages — they
don't install any files in common.

Since we're no longer pulling just one output into nativeBuildInputs
of packages using wayland-scanner, we can use upstream's
wayland-scanner.pc, which gets installed into the dev output, rather
than providing our own.

wegank commented:
The remaining occurrences of `withLibraries` are in

- pkgs/development/libraries/wayland/protocols.nix
- pkgs/tools/graphics/mesa-demos/default.nix (well...)

alyssais commented:
Things that still need to be checked:

- [ ] nix-review to catch packages that relied on wayland in `buildInputs` and `strictDeps = false` to get wayland-scanner
- [ ] test cross compilation of all packages that directly depend on wayland-scanner (they might need wayland-scanner in `depsBuildBuild` now)

alyssais committed:
clightd: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
python3.pkgs.pywayland: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

SuperSandro2000 reviewed (approved):
Thanks for cleaning this up

alyssais committed:
river-tag-overlay: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
hello-wayland: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
gst_all_1.gst-plugins-base: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

alyssais committed:
way-displays: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

alyssais committed:
gtk-layer-shell: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais commented:
Some packages require patches to make this work:

- [ ] [waffle](https://gitlab.freedesktop.org/mesa/waffle/-/merge_requests/121)
- [ ] [extra-cmake-modules](https://invent.kde.org/frameworks/extra-cmake-modules/-/merge_requests/336)

I'll keep going with trying to identify packages that will need to have their dependencies adjusted, to try to identify any more that will need fixes, but before undrafting I'll wait a bit for the upstreams to hopefully accept my patches so I can fetch them from a stable commit hash her.

SuperSandro2000 commented:
I am not sure how much this is in scope of this PR but we might want to set strictDeps for the fixed packages to avoid similar regressions in the future.

alyssais commented:
> I am not sure how much this is in scope of this PR but we might want to set strictDeps for the fixed packages to avoid similar regressions in the future.

I'm a bit scared of `strictDeps` with graphical stuff because of it causing problems with `wrapGAppsHook` (not sure if that's fixed by now), and I've already gone through about 60 packages (got a batch not pushed yet), so I'll leave that for future work.

alyssais committed:
squeekboard: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

alyssais committed:
maliit-framework: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
phoc: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
rofi-wayland: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
tofi: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
waylock: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

alyssais committed:
kitty: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
vlc: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

alyssais committed:
looking-glass-client: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
cagebreak: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

alyssais committed:
cardboard: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
gamescope: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
labwc: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
river: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

River also needs `wayland` itself in `buildInputs`.

alyssais committed:
wio: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
enlightenment.efl: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
gnome3.mutter43: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
gnome3.mutter: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
gst_all_1.gst-plugins-bad: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

alyssais committed:
libwpe-fdo: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

alyssais committed:
vulkan-cts: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
vulkan-tools: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.krfb: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
latte-dock: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
phosh: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
phosh-mobile-settings: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
deepin.dwayland: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.kscreenlocker: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.kwin: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.plasma-integration: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.plasma-remotecontrollers: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.plasma-workspace: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.xdg-desktop-portal-kde: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.kguiaddons: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.kidletime: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
plasma5Packages.kwayland: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais committed:
waffle: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

Also, add pkg-config to depsBuildBuild to fix cross.

alyssais committed:
webkitgtk: depend on wayland-scanner

Currently wayland-scanner is defined to be wayland.bin, but we want to
split wayland-scanner into a separate package.

The previous explicit -DWAYLAND_SCANNER= is removed - the binary is in
a separate bin output and it works out of the box just with
wayland-scanner in nativeBuildInputs.

alyssais committed:
fcitx5: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

alyssais commented:
Okay, I have come to the conclusion that it's not feasible to nixpkgs-review this, as there's too much unrelated broken stuff and too much churn.  But I have fixed lots of the packages that will now need to depend on wayland-scanner, by testing everything that needs wayland-protocols.  This won't be exhaustive, as some packages that build only their own Wayland protocols rather than any upstream ones won't be found, but those are hopefully rare and I don't think there's a better way to find those than just waiting for them on Hydra in staging-next.  They're extremely simple to fix when found, anyway.

I'll still check for cross-compilation regressions for all the packages I've changed here before I undraft.

alyssais committed:
clightd: fix cross

flokli commented:
@alyssais any chance you can rebase this to somewhere past #318226?

alyssais commented:
@flokli probably not any time soon — go ahead and take it over if you want.

flokli commented:
I pushed a rebase, am gonna do some testing probably tomorrow, currently still rebuilding the world before all that :-)

flokli committed:
vaapiIntel: depend on wayland-scanner

Otherwise, wayland-scanner would be picked up from the wayland in
buildInputs, which isn't cross-friendly and will stop working when we
split wayland-scanner into a separate package.

flokli committed:
freerdp: don't set WAYLAND_SCANNER explicitly

This is in fact wrong since moving the binary to the `bin` output, and
works just fine only with it in `nativeBuildInputs`.

flokli committed:
freerdp3: don't set WAYLAND_SCANNER explicitly

This is in fact wrong since moving the binary to the `bin` output, and
works just fine only with it in `nativeBuildInputs`.

Also, add wayland-scanner to buildInputs, so the pkg-config machinery
does discover it.

flokli committed:
gamescope: fix cross

This needs to be buildPackages.patchelfUnstable, as it's executed by the
build platform, during the build.

flokli commented:
Did another pass over this, applying some fixes.

I tested the build of the following packages. They all did build (native,
x86_64-linux).

I also tested cross-compilation for aarch64-linux (from x86_64-linux).

In places where cross compilation did fail, I compared with
cross status before the PR, to ensure it didn't regress due to this PR.

The checkout I compared with was origin/master
(617bfe5fa96cd753f7bfc60be43050daac390425) with
617bfe5fa96cd753f7bfc60be43050daac390425 cherry-picked on top, as I had that
worktree laying around.

 - [x] `cagebreak` ([x] cross)
 - [ ] ~`cardboard` (marked as `meta.broken`)~
 - [x] `clightd` ([x] cross)
 - [x] `deepin.dwayland` (cross already previously broken for unrelated reasons, "Could NOT find ECM (missing: ECM_DIR)")
 - [x] `enlightment.efl` (cross already previously broken on `ibus` dep)
 - [x] `fcitx5` (cross already previously broken on `xcb-imdkit` dep)
 - [x] `gamescope` ([x] cross fixed thanks to this PR, and by fixing patchelf reference)
 - [x] `gnome.mutter` (cross already previously broken on `libgweather` dep)
 - [x] `gnome.mutter43` (cross already previously broken on `libgweather` dep)
 - [x] `gst_all_1.gst-plugins-bad` ([x] cross)
 - [x] `gst_all_1.gst-plugins-base` ([x] cross)
 - [x] `gtk-layer-shell` ([x] cross)
 - [x] `hello-wayland` ([x] cross)
 - [x] `kitty` (cross already previously broken for unrelated reasons, "No package 'harfbuzz' found")
 - [x] `labwc` ([x] cross)
 - [x] `latte-dock` (cross already previously broken on `libgweather` dep)
 - [x] `libwpe-fdo` ([x] cross)
 - [x] `looking-glass-client` (x86 only according to meta)
 - [x] `maliit-framework` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `phoc` (cross already previously broken on `libgweather` dep)
 - [x] `phosh` (cross already previously broken on `libgweather` dep)
 - [x] `phosh-mobile-settings` (cross already previously broken on `libgweather` dep)
 - [x] `plasma5Packages.kguiaddons` (cross already previously broken on "qtsvg detected mismatched Qt dependencies") 
 - [x] `plasma5Packages.kidletime` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `plasma5Packages.krfb` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `plasma5Packages.kscreenlocker` (cross already previously broken on "qtsvg detected mismatched Qt dependencies") 
 - [x] `plasma5Packages.kwayland` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `plasma5Packages.kwin` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `plasma5Packages.plasma-integration` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `plasma5Packages.plasma-remotecontrollers` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `plasma5Packages.plasma-workspace` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `plasma5Packages.xdg-desktop-portal-kde` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `python3.pkgs.pywayland` ([x] cross)
 - [x] `river` (cross broken due to already previously broken zig build)
 - [x] `river-tag-overlay` ([x] cross)
 - [x] `rofi-wayland` ([x] cross)
 - [x] `squeekboard` (cross already fails with "Relocations in generic ELF (EM: 62)" during linking)
 - [x] `tofi` ([x] cross)
 - [x] `vaapiIntel` (x86 only according to meta)
 - [x] `vlc` (cross already previously broken on "qtsvg detected mismatched Qt dependencies")
 - [x] `vulkan-cts` (cross already previously broken with missing "SPIRV-Headers")
 - [x] `vulkan-tools` ([x] cross)
 - [x] `waffle` ([x] cross)
 - [x] `way-displays` ([x] cross)
 - [x] `waylock` (cross broken due to already previously broken zig build)
 - [x] `webkitgtk` ([x] cross, but only `webkitgtk_6_0` tested)
 - [x] `wio` ([x] cross)
 - [x] `wayland-scanner` ([x] cross)
 - [x] `clightd` ([x] cross)
 - [x] `freerdp` ([x] cross)
 - [x] `freerdp3` (cross build fails ("No target "sdl-common-res2bin"), but previously even its deps were broken)

Please take a look. This is ready for review, undrafting.

K900 commented:
KDE bits LGTM, the rest seems good at a glance but did not look very closely.

SomeoneSerge commented on pkgs/development/libraries/wayland/scanner.nix:
```diff
@@ -0,0 +1,48 @@
+  separateDebugInfo = true;
+
+  mesonFlags = [
+    "-Ddocumentation=false"
+    "-Dlibraries=false"
+    "-Dtests=false"
```
nit: `lib.mesonBool`

flokli replied:
Sure, applied here and in `default.nix, no rebuild.

SomeoneSerge commented on pkgs/development/libraries/wayland/scanner.nix:
```diff
@@ -0,0 +1,48 @@
+    "-Dtests=false"
+  ];
+
+  depsBuildBuild = [ pkg-config ];
```
I see this was in the original `wayland` derivation but without a comment why do we need `BuildBuild` in addition to `BuildHost` 🤔

alyssais replied:
This is required for any package that uses pkg-config to find a build dependency.  pkg-config always needs to be one step "up" from the packages it's finding.

alyssais replied:
(For this reason, if it's non-obvious and isn't already, it would make a lot more sense for this to be explained in doc/, rather than in every package that builds programs for the build system that use a library)

SomeoneSerge replied:
> This is required for any package that uses pkg-config to find a build dependency

Oh, actually, this makes sense. Thanks

flokli replied:
Hmmh, but this would be a general cross-compilation docs change, unrelated to this wayland PR, no?

SomeoneSerge replied:
> Hmmh, but this would be a general cross-compilation docs change, unrelated to this wayland PR, no?

Yes, out of scope

SomeoneSerge commented on pkgs/development/libraries/wayland/default.nix:
```diff
@@ -53,13 +41,13 @@ stdenv.mkDerivation (finalAttrs: {
     "-Ddocumentation=${lib.boolToString withDocumentation}"
-    "-Dlibraries=${lib.boolToString withLibraries}"
     "-Dtests=${lib.boolToString withTests}"
+    "-Dscanner=false"
```
Idk do we need a comment that there is a separately maintained `wayland-scanner` derivation now which is why this is not to be configurable, or is this self-evident...

flokli replied:
added a comment.

SomeoneSerge commented on pkgs/development/libraries/wayland/scanner.nix:
```diff
@@ -0,0 +1,48 @@
+    meson
+    pkg-config
+    ninja
+  ] ++ lib.optional (stdenv.hostPlatform != stdenv.buildPlatform) wayland-scanner;
```
I'm tempted to test if this is really still necessary (how come we can build native wayland-scanner without a prebuilt wayland-scanner, but the cross wayland-scanner requires a prebuilt native wayland-scanner)

Nit:

<details>

`optionals cond [ x ]` over `optional cond x` because this prevents the accidental `optional cond [ x ]`

</details>

flokli replied:
It is necessary, as it fails to otherwise fails to find `wayland-scanner` for host:

```
Run-time dependency expat found: YES 2.6.2
Run-time dependency libxml-2.0 found: YES 2.13.3
Program embed.py found: YES (/nix/store/04gg5w1s662l329a8kh9xcwyp0k64v5a-python3-3.12.4/bin/python3.12 /build/wayland-1.23.0/src/embed.py)
Found pkg-config: YES (/nix/store/blb1ci2nwrmgzb3g1xl4mvqpwkhm3s3b-pkg-config-wrapper-0.29.2/bin/pkg-config) 0.29.2
Did not find CMake 'cmake'
Found CMake: NO
Build-time dependency wayland-scanner found: NO (tried pkgconfig)

src/meson.build:84:15: ERROR: Dependency "wayland-scanner" not found, tried pkgconfig

A full log can be found at /build/wayland-1.23.0/build/meson-logs/meson-log.txt
error: build of '/nix/store/v16pnil9wkd1hhvrfp7n6sbgl86000ms-wayland-scanner-aarch64-unknown-linux-gnu-1.23.0.drv' on 'ssh://root@ws.my.flokli.io' failed: builder for '/nix/store/v16pnil9wkd1hhvrfp7n6sbgl86000ms-wayland-scanner-aarch64-unknown-linux-gnu-1.23.0.drv' failed with exit code 1
```

flokli replied:
I personally ignore that nit, if we want that it should be a tree-wide thing, because why otherwise do we have it.

SomeoneSerge replied:
😱. Thanks.

SomeoneSerge commented on pkgs/tools/graphics/vulkan-tools/default.nix:
```diff
@@ -64,14 +66,13 @@ stdenv.mkDerivation rec {
   dontPatchELF = true;
 
-  env.PKG_CONFIG_WAYLAND_SCANNER_WAYLAND_SCANNER="${buildPackages.wayland-scanner}/bin/wayland-scanner";
+  env.PKG_CONFIG_WAYLAND_SCANNER_WAYLAND_SCANNER = "${lib.getBin buildPackages.wayland-scanner}/bin/wayland-scanner";
```
This is a simple `getExe` now, but I should stop bike-shedding

flokli replied:
bike sheds addressed.

SomeoneSerge reviewed (approved):
@flokli already did more testing than I have the compute for, and code-wise this LGTM

flokli committed:
wayland-scanner: set mainProgram

flokli committed:
vulkan-tools: use lib.getExe

alyssais commented:
Thanks for picking this up.  My main concern is that there might be more packages that now depend on wayland-scanner that didn't last year, and identifying them is hard — it's why I never finished this PR — but I think as long as you're willing to actively watch Hydra during the next staging cycle to catch regressions, it should be fine.

trofi commented:
Bisect says 921bd99f8de553bd45be0ed298619589f2b7fee4 `wayland-scanner: split from wayland` broke `aquamarine` build in `staging` as:

```
$ nix build --no-link -f. -L aquamarine
...
aquamarine> Running phase: buildPhase
aquamarine> build flags: -j16 SHELL=/nix/store/srq8ja7vpm3b95p2rhw149mx9pbfaqrv-bash-5.2p32/bin/bash
aquamarine> [  3%] Generating /build/source/protocols/xdg-shell.cpp, /build/source/protocols/xdg-shell.hpp
aquamarine> [  7%] Generating /build/source/protocols/linux-dmabuf-v1.cpp, /build/source/protocols/linux-dmabuf-v1.hpp
aquamarine> [ 11%] Generating /build/source/protocols/wayland.cpp, /build/source/protocols/wayland.hpp
aquamarine> Couldn't load proto
aquamarine> make[2]: *** [CMakeFiles/aquamarine.dir/build.make:74: /build/source/protocols/wayland.cpp] Error 1
aquamarine> make[2]: *** Waiting for unfinished jobs....
aquamarine> make[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/aquamarine.dir/all] Error 2
aquamarine> make: *** [Makefile:136: all] Error 2
```

flokli commented:
Thanks for the report, I'll take a close look once I rebuilt things until there :-)

The package recipe might need some slight adjustments.

flokli commented:
It was an issue with the upstream package picking from the wrong .pc file, which we exposed with the wayland-scanner split.

Fix in https://github.com/NixOS/nixpkgs/pull/335788.

trofi commented:
Another casualty of 921bd99f8de553bd45be0ed298619589f2b7fee4 `wayland-scanner: split from wayland` is `wine64Packages.wayland` on `staging`:

```
$ nix build --no-link -f. wine64Packages.wayland
...
wine64-wayland> checking for wayland-scanner... Package wayland-scanner was not found in the pkg-config search path.
wine64-wayland> Perhaps you should add the directory containing `wayland-scanner.pc'
wine64-wayland> to the PKG_CONFIG_PATH environment variable
wine64-wayland> No package 'wayland-scanner' found
wine64-wayland> no
...
wine64-wayland> configure: error: Wayland 64-bit development files not found or not new enough, the Wayland driver won't be supported.
wine64-wayland> This is an error since --with-wayland was requested.
```

flokli commented:
Try adding wayland-scanner to buildInputs, so pkg-config can discover it?

trofi commented:
Sounds good. Proposed the addition as:
- https://github.com/NixOS/nixpkgs/pull/336001

SomeoneSerge commented:
> Try adding wayland-scanner to buildInputs, so pkg-config can discover it?

Idk much about how wine uses wayland-scanner, but why `buildInputs` and not native? Are you sure this isn't just working because strictDeps are off?

trofi commented:
wine's generic nix code is a bit unusual in how it pulls in `pkgs` due to multi-abi support. Bringing `wayland-scanner` the same way it used to be exposed sounds simplest just to make it work again.

flokli commented:
> wine's generic nix code is a bit unusual in how it pulls in `pkgs` due to multi-abi support. Bringing `wayland-scanner` the same way it used to be exposed sounds simplest just to make it work again.

Fair enough, cleaning this up can be a separate PR, that PR is not making it more messy than it was before.

trofi commented:
`hyprland` is another impacted package:

```
$ nix build --no-link -f. -L hyprland
...
hyprland> cd /build/source && hyprwayland-scanner --wayland-enums //nix/store/k3wz91y35macps9c811s68z207vxymc7-wayland-1.23.0/share/wayland/wayland.xml /build/source/protocols/
hyprland> Couldn't load proto
```

flokli commented:
> `hyprland` is another impacted package:
> 
> ```
> $ nix build --no-link -f. -L hyprland
> ...
> hyprland> cd /build/source && hyprwayland-scanner --wayland-enums //nix/store/k3wz91y35macps9c811s68z207vxymc7-wayland-1.23.0/share/wayland/wayland.xml /build/source/protocols/
> hyprland> Couldn't load proto
> ```

This is the same issue as it was in https://github.com/hyprwm/aquamarine/pull/55. I sent a PR at https://github.com/hyprwm/Hyprland/pull/7467.
